src/ngraph/ngraph_emitter.cc

```diff
@@ -1285,6 +1286,119 @@ void Emitter::CreateLayerOps() {
           data, ngraph::AxisSet{static_cast<size_t>(seq_axis)});
     }
   };
+  ngraph_op_funcs_["SoftmaxOutput"] = [this](const NodePtr& node) {
+    auto input = op_map_[node->inputs_[0]];
+    auto in_shape = input->get_shape();
+
+    ngraph::AxisSet axes;
+    if (get_default(node, "multi_output", false)) {
+      axes.insert(1);
+    } else if (get_default(node, "preserve_shape", false)) {
+      axes.insert(in_shape.size() - 1);
+    } else {
+      for (size_t i = 1; i < in_shape.size(); ++i) {
+        axes.insert(i);
+      }
+    }
+    return std::make_shared<ngraph::op::Softmax>(input, axes);
+  };
+}
+
+void Emitter::CreateLossOps() {
+  // These functions are in place to provide a mechanism for manually specifying
+  // backpropgation methods for Loss functions. We do this because MXNet
+  // provides a number of options that only effect the output of backprop, not
+  // forward prop, and are difficult or impossible to integrate into
+  // nGraph's autodiff functionality
+  loss_op_backward_funcs_["SoftmaxOutput"] = [this](
+      const NodePtr& node, const NgraphNodePtr& adjoint) {
+    float grad_scale = get_default(node, "grad_scale", 1.0f);
+    float ignore_label = get_default(node, "ignore_label", -1.0f);
+    float smooth_alpha = get_default(node, "smooth_alpha", 0.0f);
+
+    bool use_ignore = get_default(node, "use_ignore", false);
+    bool out_grad = get_default(node, "out_grad", false);
+
+    std::string norm = get_default(node, "normalization", std::string("null"));
+
+    auto softmax = op_map_[node];
+    auto label = op_map_[node->inputs_[1]];
+    bool ignore = false;
+    NgraphNodePtr mask;
+
+    if (label->get_shape() != softmax->get_shape()) {
+      if (use_ignore) {
+        ignore = true;
+        mask = cast_result(
+            std::make_shared<ngraph::op::NotEqual>(
+                label, makeConstant(node, std::to_string(ignore_label))),
+            getType(node->dtype_));
+      }
+      size_t axis = op_map_[node->inputs_[0]]->get_shape().size() - 1;
+      if (get_default(node, "multi_output", false)) {
+        axis = 1;
+      }
+      label = std::make_shared<ngraph::op::OneHot>(label, softmax->get_shape(),
+                                                   axis);
+    }
+
+    if (smooth_alpha != 0.0f) {
+      int num_classes = 1;
+      auto shape = softmax->get_shape();
+      if (get_default(node, "multi_output", false)) {
+        num_classes = shape[1];
+      } else if (get_default(node, "preserve_shape", false)) {
+        num_classes = shape.back();
+      } else {
+        for (size_t i = 1; i < shape.size(); ++i) {
+          num_classes *= shape[i];
+        }
+      }
+      auto one = makeConstant(node, "1");
```

Comment:
Since `makeContant(...)` now supports non-string types, might it make sense to use `makeConstant(node, 1)` instead?

IMO presenting the value as a string here, and requiring `makeConstant(...)` to parse it as an integer at runtime, seems like unnecessary complexity.  And the unsuspecting reader of the code might feel the need to dig into the question of why a string is used here, since it seems unnecessary.

This comment applies to the `makeConstant(...)` calls below as well.

Reply:
I don't like converting things between finite precision formats, thus the string.